### PR TITLE
sql: show create all schemas command

### LINKED ISSUE #70620: Include custom schema definitions in SHOW CREATE ALL TABLES
Since `cockroach dump` is being deprecated, I'm trying to use `SHOW CREATE ALL TABLES` to obtain a schema dump that I can check into version control so we have diffs of DB schema changes and a starting point for running migrations. However, we use user defined schemas, and these are not included in the output.

**Describe the solution you'd like**
I'd like that `SHOW CREATE ALL TABLES` outputs DDL for user defined schemas (`CREATE SCHEMA` statements), so that the complete output can be used to create a new database with the correct schema (for integration tests, test environments, whatnot).

**Describe alternatives you've considered**
I could manually maintain an additional file with user defined schema definitions, but it would be better if it could all be managed by this command.

### PR DESCRIPTION
**Pull request by jackcwu**
Release note (sql change): SHOW CREATE ALL SCHEMAS is a command that allows the user to get the statements to recreate the schemas of the current database. The command returns a flat log of the create statements for schemas.

Example output:
```
SHOW CREATE ALL SCHEMAS;
----
CREATE SCHEMA test;
CREATE SCHEMA test2;
```

Resolves: https://github.com/cockroachdb/cockroach/issues/70620

### CODE REVIEW

File: pkg/sql/sem/builtins/generator_builtins.go [hunk 1]
```diff
@@ -1879,9 +1896,10 @@ func (p *payloadsForTraceGenerator) Close(_ context.Context) {
 }
 
 var showCreateAllTablesGeneratorType = types.String
+var showCreateAllSchemasGeneratorType = types.String
 
 // Phase is used to determine if CREATE statements or ALTER statements
-// are being generated for showCreateAllTables.
+// are being generated for showCreateAllTables or showCreateAllSchemas.
```
Comment (RichardJCai) [inline pkg/sql/sem/builtins/generator_builtins.go:1902]: We don't need to use Phase for `showCreateAllSchemas` only one "phase" which is to output create statements.

File: pkg/sql/sem/builtins/generator_builtins.go [hunk 2]
```diff
@@ -402,6 +402,23 @@ in the alter statements.
...
+			Class: tree.GeneratorClass,
+		},
+		makeGeneratorOverload(
+			tree.ArgTypes{
+				{"database_name", types.String},
+			},
+			showCreateAllSchemasGeneratorType,
+			makeShowCreateAllSchemasGenerator,
+			`Returns rows of CREATE schema statements followed by 
+ALTER schema statements that add schema constraints.
```
Comment (RichardJCai) [inline pkg/sql/sem/builtins/generator_builtins.go:420]: No alter schema statements for schema.

**Review by RichardJCai — CHANGES_REQUESTED** [review]
Looks pretty good overall, some small comments. Well done overrall

The failing logic tests you can simply -rewrite

See failing TestStaticCheck though, there are some unused functions.
https://teamcity.cockroachdb.com/viewLog.html?buildId=3540682&buildTypeId=Cockroach_UnitTests#testNameId7798675035218714145

File: pkg/sql/sem/builtins/generator_builtins.go (hunk 1 above)
Comment (jackcwu) [reply pkg/sql/sem/builtins/generator_builtins.go]: Fixed

File: pkg/sql/sem/builtins/generator_builtins.go (hunk 2 above)
Comment (jackcwu) [reply pkg/sql/sem/builtins/generator_builtins.go]: Fixed

**Review by jackcwu — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71138#-:-MlGsuA5BivKkYL3hjGR:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rafiss and @RichardJCai)





Acknowledged

**Review by RichardJCai — CHANGES_REQUESTED** [review]
```
sql: show create all schemas command
Release note: None

Resolves: #70620
```

The release note should not be None here, see the release note for the show create all tables for example.

> Reviewed 4 of 14 files at r1, 3 of 10 files at r2, 21 of 22 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71138#-:-MlRK1My6qpcgUcSjdEm:btxyoz4)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jackcwu, @rafiss, and @RichardJCai)

___
*[pkg/ccl/logictestccl/testdata/logic_test/placement, line 44 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71138#-MlRJG1z-halvrzfXfUR:-MlRJG1z-halvrzfXfUS:b-85bp) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9eecf871b594609b2eace35e4451b3db192da01/pkg/ccl/logictestccl/testdata/logic_test/placement#L44)):*
> ```
> statement error a region-survivable database cannot also have a restricted placement policy
> ALTER DATABASE region_survivable_default PLACEMENT RESTRICTED
> ```

Why did this change?
___
*[pkg/sql/crdb_internal.go, line 2284 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71138#-MlRJLiL3zEgcsTlFsFd:-MlRJLiL3zEgcsTlFsFe:bqbarlz) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0648f2f12c0e6dce9f436896078440f8aafb59c/pkg/sql/crdb_internal.go#L2284)):*
> ```Go
> 
> var crdbInternalCreateSchemaStmtsTable = virtualSchemaTable{
> 	comment: "CREATE statements for all user defined types accessible by the current user in current database (KV scan)",
> ```

s/user defined types/user defined schemas
___
*[pkg/sql/parser/sql.y, line 5662 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71138#-MlRJo9-1K87eDEOgEJu:-MlRJo9-1K87eDEOgEJv:b-l6qzmh) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0648f2f12c0e6dce9f436896078440f8aafb59c/pkg/sql/parser/sql.y#L5662)):*
> ```Yacc
> | SHOW CREATE ALL SCHEMAS
> 	{
> 		$$.val = &tree.ShowCreateAllSchemas{}
> ```

Nit: use spaces instead of tabs here

**Review by jackcwu — COMMENTED** [review]
fixed

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71138#-:-MlRYk2K0rkVRfO2tjL7:b-7ql1td)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rafiss and @RichardJCai)

___
*[pkg/ccl/logictestccl/testdata/logic_test/placement, line 44 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71138#-MlRJG1z-halvrzfXfUR:-MlRYm6YEAuFXatDNZCI:b6yxc53) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9eecf871b594609b2eace35e4451b3db192da01/pkg/ccl/logictestccl/testdata/logic_test/placement#L44)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Why did this change?
</blockquote></details>

happened on the rewrite most likely
___
*[pkg/sql/crdb_internal.go, line 2284 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71138#-MlRJLiL3zEgcsTlFsFd:-MlRYuPN4xuUZfqnTBQn:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0648f2f12c0e6dce9f436896078440f8aafb59c/pkg/sql/crdb_internal.go#L2284)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

s/user defined types/user defined schemas
</blockquote></details>

fixed
___
*[pkg/sql/parser/sql.y, line 5662 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71138#-MlRJo9-1K87eDEOgEJu:-MlRYvDRACNOMmLpBc3H:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0648f2f12c0e6dce9f436896078440f8aafb59c/pkg/sql/parser/sql.y#L5662)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Nit: use spaces instead of tabs here
</blockquote></details>

fixed

**Review by RichardJCai — CHANGES_REQUESTED** [review]
I would change the release note to:

Release note (sql change): SHOW CREATE ALL SCHEMAS is a command that allows the user to get the statements to recreate the schemas of the current database. The command returns a flat log of the create statements for schemas.
Example output:

SHOW CREATE ALL SCHEMAS;
----
CREATE SCHEMA test;
CREATE SCHEMA test2;

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71138#-:-MlW2G0SEqYWGtDD1v97:bif2yd2)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jackcwu, @rafiss, and @RichardJCai)

___
*[pkg/sql/crdb_internal.go, line 2284 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71138#-MlRJLiL3zEgcsTlFsFd:-MlW22Oz-NqS8O1aGY3U:b-lnh04k) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0648f2f12c0e6dce9f436896078440f8aafb59c/pkg/sql/crdb_internal.go#L2284)):*
<details><summary><i>Previously, jackcwu wrote…</i></summary><blockquote>

fixed
</blockquote></details>

This doesn't look like it was updated.
___
*[pkg/sql/sem/builtins/generator_builtins.go, line 1924 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/71138#-MlW2Aak1oB45Kij6psD:-MlW2Aak1oB45Kij6psE:b-pkwxod) ([raw file](https://github.com/cockroachdb/cockroach/blob/3eebc39433cf601895f370a4a34fe43e0906b2c3/pkg/sql/sem/builtins/generator_builtins.go#L1924)):*
> ```Go
> 	curr  tree.Datum
> 	idx   int
> 	phase Phase
> ```

Shouldn't need this.

**Review by jackcwu — COMMENTED** [review]
Fixed

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71138#-:-MlW6rwb48L1fhBCwYBK:b-886gsx)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rafiss and @RichardJCai)

___
*[pkg/sql/crdb_internal.go, line 2284 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71138#-MlRJLiL3zEgcsTlFsFd:-MlW6tqE8tUqeiSfsCl9:blscicd) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0648f2f12c0e6dce9f436896078440f8aafb59c/pkg/sql/crdb_internal.go#L2284)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

This doesn't look like it was updated.
</blockquote></details>

Oh my bad, I accidentally changed this for the virtual table for type statements. Should be fixed now
___
*[pkg/sql/sem/builtins/generator_builtins.go, line 1924 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/71138#-MlW2Aak1oB45Kij6psD:-MlW7-IN5lel6RNbF-4T:b-7qtyx) ([raw file](https://github.com/cockroachdb/cockroach/blob/3eebc39433cf601895f370a4a34fe43e0906b2c3/pkg/sql/sem/builtins/generator_builtins.go#L1924)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Shouldn't need this.
</blockquote></details>

Removed the phase

**Review by RichardJCai — APPROVED** [review]
Nice! LGTM

You can bors it when CI passes.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71138#-:-MlW8O6QEyhXURTZlFKQ:b7658s2)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rafiss and @RichardJCai)

**Pull request merged**